src/OrchardCore/OrchardCore.DisplayManagement.Liquid/RequestLiquidEventHandler.cs

```diff
@@ -0,0 +1,80 @@
+using System;
+using System.Threading.Tasks;
+using Fluid;
+using Fluid.Values;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Internal;
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Liquid;
+
+namespace OrchardCore.DisplayManagement.Liquid
+{
+    /// <summary>
+    /// Provides access to <see cref="HttpRequest"/> properties if the template is running in
+    /// a web request.
+    /// </summary>
+    public class RequestLiquidTemplateEventHandler : ILiquidTemplateEventHandler
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private HttpContext _httpContext;
+
+        static RequestLiquidTemplateEventHandler()
+        {
+            TemplateContext.GlobalMemberAccessStrategy.Register<HttpRequest>((request, name) =>
+            {
+                switch (name)
+                {
+                    case "QueryString": return new StringValue(request.QueryString.Value);
+                    case "ContentType": return new StringValue(request.ContentType);
+                    case "ContentLength": return new NumberValue(request.ContentLength ?? 0);
+                    case "Cookies": return new ObjectValue(request.Cookies);
+                    case "Headers": return new ObjectValue(new HeaderDictionaryWrapper(request.Headers));
+                    case "Query": return new ObjectValue(request.Query);
+                    case "Form": return new ObjectValue(request.Form);
+                    case "Protocol": return new StringValue(request.Protocol);
+                    case "Path": return new StringValue(request.Path.Value);
+                    case "PathBase": return new StringValue(request.PathBase.Value);
+                    case "Host": return new StringValue(request.Host.Value);
+                    case "IsHttps": return new BooleanValue(request.IsHttps);
+                    case "Scheme": return new StringValue(request.Scheme);
+                    case "Method": return new StringValue(request.Method);
+
+                    default: return null;
+                }
+            });
+
+            TemplateContext.GlobalMemberAccessStrategy.Register<RequestCookieCollection>((cookies, name) => cookies[name]);
+            TemplateContext.GlobalMemberAccessStrategy.Register<QueryCollection>((queries, name) => queries[name].ToArray());
+            TemplateContext.GlobalMemberAccessStrategy.Register<FormCollection>((forms, name) => forms[name].ToArray());
+            TemplateContext.GlobalMemberAccessStrategy.Register<HeaderDictionaryWrapper>((headers, name) => headers.HeaderDictionary[name].ToArray());
+        }
```

Comment:
With cookies you don't use `.ToArray()`.

But you use it with others => returns a string[].

Reply:
Oh yes there are `StringValues`